### PR TITLE
Vice Access request and errors

### PR DESCRIPTION
**Pull request by sri2k1us**
Limit Reached:
image

Request VICE Access:
image

Account Banned:
image

P.S: Admin feature to list and process VICE Access request will be in a separate PR.

### CODE REVIEW

File: src/components/apps/AppName.js [hunk 1]
```diff
@@ -40,6 +100,50 @@ function AppName(props) {
                 />
             </Link>
         );
+    } else if (runErrorCodes?.length > 0) {
+        if (runErrorCodes[0] === ERROR_CODES.ERR_PERMISSION_NEEDED) {
```
Comment (ash3rz) [inline src/components/apps/AppName.js:104]: Maybe `const runErrorCode = runErrorCodes[0];` could be added for some reuse?

File: src/components/apps/AppName.js [hunk 2]
```diff
@@ -40,6 +100,50 @@ function AppName(props) {
...
+                        component="button"
+                        href="#"
+                        onClick={() => setAccessRequestDialogOpen(true)}
+                        color="primary"
+                    >
+                        {name}
+                    </MuiLink>
+                    <AccessRequestDialog
+                        open={accessRequestDialogOpen}
+                        baseId={baseDebugId}
```
Comment (ash3rz) [inline src/components/apps/AppName.js:117]: This dialog and the `RunErrorDialog` should have their own unique IDs

File: src/components/apps/launch/index.js [hunk 3]
```diff
@@ -78,31 +89,68 @@ const Launch = ({ app, launchError, loading }) => {
...
+                    <Button
+                        onClick={() => setAccessRequestDialogOpen(true)}
+                        color="primary"
+                        variant="contained"
+                    >
+                        {t("requestAccess")}
+                    </Button>
+                    <AccessRequestDialog
+                        open={accessRequestDialogOpen}
+                        baseId={baseId}
```
Comment (ash3rz) [inline src/components/apps/launch/index.js:121]: Same thing here with having a unique ID

File: src/components/apps/launch/index.js [hunk 4]
```diff
@@ -78,31 +89,68 @@ const Launch = ({ app, launchError, loading }) => {
...
-            submitAnalysis={(submission, onSuccess, onError) => {
-                setSubmissionError(null);
-                submitAnalysisMutation({ submission, onSuccess, onError });
-            }}
-            saveQuickLaunch={(quickLaunch, onSuccess, onError) => {
-                setSubmissionError(null);
-                addQuickLaunchMutation({ quickLaunch, onSuccess, onError });
-            }}
-        />
+        <>
```
Comment (ash3rz) [inline src/components/apps/launch/index.js:133]: This extra fragment doesn't look to be necessary

File: public/static/locales/en/vice.json [hunk 5]
```diff
@@ -0,0 +1,9 @@
+{
+    "accessRequestTitle": "Request access to VICE analyses (e.g: Jupyter, RStudio etc...)",
```
Comment (ash3rz) [inline public/static/locales/en/vice.json:2]: Picking a random spot - for these messages maybe we should pick a particular word to use.  Here we're saying `VICE analyses`, in another message it says `... not permitted to launch VICE apps`, and in one other it says `... currently running VICE jobs`.  We've had some discussions in Slack about how some users are confused by "analyses" in regards to VICE, so it's stuck in my head as something to think about 😄

**Review by ash3rz — APPROVED** [review]
LGTM :shipit: 
I just had some minor comments and a suggestion about the wording we're using in these dialogs.  From the screenshots, it looks like the `contact support` link is in a smaller font again compared to the rest.

File: src/components/vice/AccessRequestDialog.js [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+import requestAccess from "serviceFacades/vice/accessRequest";
+import constants from "../../constants";
+
+import DEDialog from "components/utils/DEDialog";
+import { useConfig } from "contexts/config";
+import { Button, CircularProgress, Typography } from "@material-ui/core";
+import ErrorTypographyWithDialog from "components/utils/error/ErrorTypographyWithDialog";
+
+const MIN_USE_CASE_LENGTH = 60;
```
Comment (psarando) [inline src/components/vice/AccessRequestDialog.js:21]: This should probably come from configs, so we (and 3rd party installs) can change it on the fly.

**Review by psarando — APPROVED** [review]
LGTM 👍 

I only had one suggestion in addition to what @ash3rz already mentioned.

File: public/static/locales/en/vice.json (hunk 5 above)
Comment (sri2k1us) [reply public/static/locales/en/vice.json:2]: Good catch. You are right, we should be consistent in our messaging.

File: src/components/vice/AccessRequestDialog.js (hunk 6 above)
Comment (sri2k1us) [reply src/components/vice/AccessRequestDialog.js:21]: I can easily add this to configs but I don't see how this value will be changed frequently or vary in different environments / 3rd party installs 🤔 (unlike `concurrent job limit` which needs to be tweaked for resource constraints in different environment / installs) .  As a matter of fact, we don't put other form field validations rules in config.

Comment (psarando) [reply src/components/vice/AccessRequestDialog.js:21]: Specifically this value seems somewhat arbitrary to me, and it's a value only high enough to force users asking for VICE access to give us a valid science use case for that access.

I can easily see this gatekeeping value not being high enough to filter out invalid use cases, and we may be asked to increase it. I can also easily see users complaining about a value that is too high, then we'll be asked to reduce it. I can also imagine a 3rd party install not wanting to require this text at all, and they'll want this to be `0`.

It seems a bit heavy to force a rebuild of the software for an arbitrary value like this, which could be changed on the fly by a config setting instead.

Comment (sri2k1us) [reply src/components/vice/AccessRequestDialog.js]: I added this to config. IMO managing our growing number of configs is more painful than this occasional / rare builds. Also don't we do builds on a config change ? Asking for my own knowledge.

**Comment by sri2k1us** [comment]
> LGTM :shipit:
> I just had some minor comments and a suggestion about the wording we're using in these dialogs. From the screenshots, it looks like the `contact support` link is in a smaller font again compared to the rest.

The `contact support` link is not inside `Typography` and do so breaks the `Link` 😩

**Comment by sri2k1us** [comment]
Thanks for the reviews. I think I have addressed all the PR comments.

File: src/components/vice/AccessRequestDialog.js (hunk 6 above)
Comment (psarando) [reply src/components/vice/AccessRequestDialog.js]: That's a good question, and I'm not sure 😬 
If so, I would think that defeats the purpose of configs 🤔

**Pull request merged**